mission_control/include/mission_control/behavior.h

```diff
@@ -32,370 +32,77 @@
   }
 
-  int WaitForExecutionTimeSlot()
-  {
-    int retval = 0;  // 0 means success
-    boost::posix_time::ptime cur_ptime;
-    boost::posix_time::ptime calc_ptime;
-    cur_ptime = boost::posix_time::second_clock::universal_time();
-    TimeStamp when = getWhen();
-
-    switch (when.type)
-    {
-      case TimeStamp::REL_TIME:
-        if (when.duration.total_seconds() == 0) break;  // now
-        boost::this_thread::sleep(when.duration);
-        break;
-      case TimeStamp::ABS_TIME:
-        // Construct the appropriate absolute datetime
-        calc_ptime = boost::posix_time::ptime(cur_ptime.date(), when.time.time_of_day());
-        // Make sure to handle if the time is less than the current time (occurs next day)
-        if (calc_ptime.time_of_day() < cur_ptime.time_of_day())
-          calc_ptime += boost::gregorian::days(1);
-        boost::this_thread::sleep(boost::system_time(calc_ptime));
-        break;
-      case TimeStamp::ABS_DATETIME:
-        if (when.time < cur_ptime)
-        {
-          ROS_ERROR("Behavior occurs in the past; aborting mission");
-          retval = -1;
-          break;
-        }
-        boost::this_thread::sleep(boost::system_time(when.time));
-        break;
-    };
-    return retval;
-  }
+  ~Behavior() {}
 
-  int computeExecutionTimeForBehavior()
+  // You must override the virtual function tick()
+  BT::NodeStatus tick() override
   {
-    int nsec = -1;
-
-    TimeStamp tout = getTimeout();
-    boost::posix_time::ptime calc_ptime,
-        cur_ptime = boost::posix_time::second_clock::universal_time();
-
-    switch (tout.type)
+    switch (status())
     {
-      case TimeStamp::REL_TIME:
-        if (tout.duration.total_seconds() == 0) return 0;  // execution time is 0 seconds
-        nsec = tout.duration.total_seconds();
-        break;
-      case TimeStamp::ABS_TIME:
-        // Construct the appropriate absolute datetime
-        calc_ptime = boost::posix_time::ptime(cur_ptime.date(), tout.time.time_of_day());
-        // Make sure to handle if the time is less than the current time (occurs next day)
-        if (calc_ptime.time_of_day() < cur_ptime.time_of_day())
-          calc_ptime += boost::gregorian::days(1);
-        nsec = (calc_ptime - cur_ptime).total_seconds();
+      case BT::NodeStatus::IDLE:
+        setStatus(BT::NodeStatus::RUNNING);
         break;
-      case TimeStamp::ABS_DATETIME:
-        if (tout.time < cur_ptime)
-        {
-          ROS_ERROR("Timeout occurs in the past; aborting mission");
-          break;
-        }
-        nsec = (tout.time - cur_ptime).total_seconds();
+      case BT::NodeStatus::RUNNING:
+        setStatus(getBehaviorStatus());
         break;
-    };
-    return nsec;
-  }
-
-  bool startBehavior()
-  {
-    std::string xml_tag = getXmlTag();
-    switch (getType())
-    {
-      case BEHAVIOR_TYPE_MSG:
-        break;
-      case BEHAVIOR_TYPE_SRV:
+      case BT::NodeStatus::FAILURE:
+        abortBehavior();
+        halt();
```

Comment:
@patapufate hmm, why not letting the action return `FAILURE`?

Reply:
Fixed. The aborting method will be implemented in the mission_control